schema/interface_definition.rnc

```diff
@@ -184,3 +187,15 @@ completionHelp = element completionHelp
     (element path { text })* &
     (element script { text })*
 }
+
+
+# childRequirements tags is a declarative way to configure basic
+# requirements of node or tagnode children.
+childRequirements = element childRequirements {
+    (element require { child+ } )? &
+    (element conflict { nodeNameAttr, child+ })* &
```

Comment:
One issue I see is that `nodeNameAttr` is now assumed to always be a single word, but conflicts are often between children at different levels 

My feeling is that there are actually two different situations we need to handle, if we want to generate helpful error messages.

The first one is a simple conflict when two nodes cannot appear together. For example, `route-reflector-client` and `route-server-client` in BGP — a neighbor cannot be both, under any circumstances. We may use something like `<mutuallyExclusiveChildren>` for that. In that case if both are found, we can show a message like `$nodeA and $nodeB cannot be configured at the same time` if any nodes from the list are found.

The second situation is a context-dependent conflict. For example, in OpenVPN, the `server` node and any of its children like `server topology` conflict with `mode client` and `mode site-to-site`. It's possible for nodes to conflict with nodes in general or with specific valued of leaf nodes.

If we find a way to handle both, that may indeed limit the need for in-script validation to ultra-context-dependent things like `route-reflector-client` allowed only if the local AS is the same as the remote AS (i.e., the peer is an iBGP peer).

The way that you are proposing will lead to a combinatorial explosion when there are more than two conflicting options.

Reply:
> The first one is a simple conflict when two nodes cannot appear together. For example, `route-reflector-client` and `route-server-client` in BGP — a neighbor cannot be both, under any circumstances. We may use something like `<mutuallyExclusiveChildren>` for that. In that case if both are found, we can show a message like `$nodeA and $nodeB cannot be configured at the same time` if any nodes from the list are found.

It would not cover cases where there are 3 children X, Y and Z where X can not be configured with Y or Z, but Y and X can both be configured as long as X is not.
This might however be less useful that what you suggest having a simple list of all mutually exclusive / conflicting children.
I have no issue with reworking `(element conflict { nodeNameAttr, child+ })* &` to be `(element mutuallyExclusiveChildren { child+ } )? &` if that is preferred by those of you that know the system and its needs better than me.

> One issue I see is that `nodeNameAttr` is now assumed to always be a single word, but conflicts are often between children at different levels
> [...]
> The second situation is a context-dependent conflict. For example, in OpenVPN, the `server` node and any of its children like `server topology` conflict with `mode client` and `mode site-to-site`. It's possible for nodes to conflict with nodes in general or with specific valued of leaf nodes.
> 
> If we find a way to handle both, that may indeed limit the need for in-script validation to ultra-context-dependent things like `route-reflector-client` allowed only if the local AS is the same as the remote AS (i.e., the peer is an iBGP peer).

While what I have here does not currently support this level of logic I am willing to take a stab at creating the required logic if you believe it will be a useful feature to have declarative configurations for conflicting descendants of different levels. The reason I did not add it is because my use-case will not benefit from it, but I would be happy to add it.

Does these changes sounds reasonable to accept?